examples/trusted-gateway-token/contracts/contracts/Token.sol

```diff
@@ -0,0 +1,122 @@
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "hardhat/console.sol";
+/**
+ * @title Token
+ * @dev Very simple ERC20 Token example, where all tokens are pre-assigned to the creator.
+ * Note they can later distribute these tokens as they wish using `transfer` and other
+ * `ERC20` functions.
+ * Based on https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v2.5.1/contracts/examples/SimpleToken.sol
+ */
+contract Token is ERC20, Ownable {
+  string public url;
+  address private _signer;
+  mapping(address=>bool) claimed;
+  error OffchainLookup(string url, bytes prefix);
+  struct BalanceProof {
+    bytes signature;
+    uint balance;
+  }
+
+  /**
+    * @dev Constructor that gives msg.sender all of existing tokens.
+    */
+  constructor(
+      string memory name,
+      string memory symbol,
+      uint256 initialSupply
+  ) ERC20(name, symbol) {
+    _mint(msg.sender, initialSupply);
+  }
+
+  function setUrl(string memory url_) external onlyOwner{
+    url = url_;
+  }
+
+  function setSigner(address signer_) external onlyOwner{
+    _signer = signer_;
+  }
+
+  function getSigner() external view returns(address){
+    return _signer;
+  }
+
+  function balanceOf(address addr) public override view returns(uint balance) {
+    if(claimed[addr]){
+      return super.balanceOf(addr);
+    }else{
+      revert OffchainLookup(url,
+        // Selector does not seem to work against public function
+        // abi.encodeWithSelector(Token.balanceOfWithProof.selector, addr)
+        abi.encodeWithSignature("balanceOfWithProof(address addr, BalanceProof memory proof)", addr)
+      );
+    }
+  }
+
+  function transfer(address recipient, uint256 amount) public override returns (bool) {
+    if(claimed[msg.sender]){
+      _transfer(msg.sender, recipient, amount);
+    }else{
+      revert OffchainLookup(url, abi.encodeWithSelector(Token.transferWithProof.selector, recipient, amount));
+    }
+    return true;
+  }
+
+  function balanceOfWithProof(address addr, BalanceProof memory proof) public view returns(uint) {
+    uint balance = super.balanceOf(addr);
+    if(claimed[addr]){
+      return balance;
+    }else{
+      console.log("balance %s", proof.balance );
+      console.log("addr %s", addr );
+      // console.log( proof.signature );
+      address recovered = recoverAddress(
+        keccak256(abi.encodePacked(proof.balance, addr)),
+        proof.signature
+      );
+      console.log("_signer %s", _signer );
+      console.log("recovered %s", recovered );
+      require(_signer == recovered, "Signer is not the signer of the token");
+      return balance + proof.balance;
+    }
+  }
+
+  function transferWithProof(address recipient, uint256 amount, BalanceProof memory proof) external returns(bool) {
+    uint l1Balance = super.balanceOf(msg.sender);
+    uint balance = balanceOfWithProof(msg.sender, proof);
+    uint diff = balance - l1Balance;
+    if(diff > 0){
+      claimed[msg.sender] = true;
```

Comment:
Probably want to set this regardless of whether there were any tokens to claim - so that the durin call is only necessary on the first transfer.

Reply:
What happens if user 0x1 initially had no token to claim offchain but happen to have the token on L1 (he just transfers), then the server updates the flat file to add the user?

Reply:
I think we're already assuming that the airdrop is fixed.